File: 24/main.java
/**
*
*	24. Swap Nodes in Pairs
*
**/


/*
Approach is similar to the python one. Refer to the python file for complexity
*/

public class Solution {
    public ListNode swapPairs(ListNode head) {
        if ((head == null)||(head.next == null))
            return head;
        ListNode second = head.next;
        head.next = swapPairs(head.next.next);
        second.next = head;
        return second;
    }
}

File: 24/main.py
####################################################
########	24. Swap Nodes in Pairs	############
####################################################



'''
This problem can be recursively solved. Go recursively down to the end and build bottom up while keeping tab on every second node
Time Complexity:O(n) every node is visited once
Space Complexity:O(1) As we keep constant tabs`
'''

class Solution:
    def swapPairs(self, head):
        if not head or not head.next:
            return head
        second = head.next
        head.next = self.swapPairs(second.next)
        second.next = head
        return second
